gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveDataset.java

```diff
@@ -0,0 +1,261 @@
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.data.management.copy.hive;
+
+import lombok.Data;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Properties;
+
+import org.apache.commons.pool2.impl.GenericObjectPool;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.Partition;
+import org.apache.hadoop.hive.metastore.api.StorageDescriptor;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.thrift.TException;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Splitter;
+import com.google.common.collect.Lists;
+import com.google.gson.Gson;
+
+import gobblin.data.management.copy.CopyConfiguration;
+import gobblin.data.management.copy.CopyableDataset;
+import gobblin.data.management.copy.CopyableFile;
+import gobblin.util.AutoReturnableObject;
+import gobblin.util.PathUtils;
+import gobblin.util.filters.HiddenFilter;
+
+
+/**
+ * Hive dataset implementing {@link CopyableDataset}.
+ */
+public class HiveDataset implements CopyableDataset {
+
+  public static final String COPY_TARGET_TABLE_ROOT =
+      HiveDatasetFinder.HIVE_DATASET_PREFIX + ".copy.target.table.root";
+  public static final String RELOCATE_DATA_FILES_KEY =
+      HiveDatasetFinder.HIVE_DATASET_PREFIX + ".copy.relocate.data.files";
+  public static final String DEFAULT_RELOCATE_DATA_FILES = Boolean.toString(false);
+
+  private static final Gson gson = new Gson();
+  private static final String databaseToken = "$DB";
+  private static final String tableToken = "$TABLE";
+
+  private final Properties properties;
+  private final FileSystem fs;
+  private final GenericObjectPool<IMetaStoreClient> clientPool;
+  private final Table table;
+  private final Splitter splitter = Splitter.on(",").omitEmptyStrings();
+
+  private final List<Path> tableLocations;
+  // Only set if table has exactly one location
+  private final Optional<Path> tableRootPath;
+  private final Optional<Path> targetTableRoot;
+  private final boolean relocateDataFiles;
+
+  public HiveDataset(FileSystem fs, GenericObjectPool<IMetaStoreClient> clientPool, Table table, Properties properties) {
+    this.fs = fs;
+    this.clientPool = clientPool;
+    this.table = table;
+    this.properties = properties;
+
+    this.tableLocations = parseLocationIntoPaths(this.table.getSd().getLocation());
+
+    this.tableRootPath = this.tableLocations.size() == 1 ?
```

Comment:
Because that path may not exist. In principle, a Hive table could contain files in different file systems even.